packages/babel-core/src/config/partial.js

```diff
@@ -6,9 +6,43 @@ import { mergeOptions } from "./util";
 import { createItemFromDescriptor } from "./item";
 import { buildRootChain, type ConfigContext } from "./config-chain";
 import { getEnv } from "./helpers/environment";
-import { validate, type ValidatedOptions } from "./validation/options";
+import {
+  validate,
+  type ValidatedOptions,
+  type RootMode,
+} from "./validation/options";
+
+import { findConfigUpwards, type ConfigFile, type IgnoreFile } from "./files";
+
+function resolveRootMode(rootDir: string, rootMode: RootMode): string {
+  switch (rootMode) {
+    case "root":
+      return rootDir;
+
+    case "upward-optional": {
+      const upwardRootDir = findConfigUpwards(rootDir);
+      return upwardRootDir === null ? rootDir : upwardRootDir;
+    }
 
-import type { ConfigFile, IgnoreFile } from "./files";
+    case "upward": {
+      const upwardRootDir = findConfigUpwards(rootDir);
+      if (upwardRootDir !== null) return upwardRootDir;
+
+      throw Object.assign(
+        (new Error(
+          `Babel was run with rootMode:"upward" but a root could not ` +
+            `be found when searching upward from "${rootDir}"`,
+        ): any),
+        {
+          code: "BABEL_ROOT_NOT_FOUND",
+          dirname: rootDir,
+        },
+      );
+    }
+    default:
+      throw new Error(`Assertion failure - unknown rootMode value`);
```

Comment:
We could also link to the website to learn more.

Reply:
I don't generally do that with assertions. This is mostly here to satisfy Flow. The user should never see this and if we got here, it means Flow didn't catch a type error somewhere, since it would mean we got a `RootMode` type that didn't match the type definition.

Reply:
Got you, there's a config validation before that anyway (`assertRootMode`).

Just FYI I used to use https://github.com/xtuc/mamacro#examples for only assertion, they are quite close to C ones.